File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])
        count = 0

        def bfs(row,col):
            queue = deque([(row,col)])
            indices = [(0,1),(0,-1),(1,0),(-1,0)]
            while queue:
                r,c = queue.popleft()
                grid[r][c] == "0"
                for dr,dc in indices:
                    nr,nc = r+dr,c+dc
                    if 0<=nr<ROWS and 0<=nc<COLS and grid[nr][nc] == "1":
                        queue.append((nr,nc))
                        grid[nr][nc] = "0"

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1":
                    count+=1
                    bfs(r,c)
        return count
        

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;
        int ROWS = grid.length;
        int COLS = grid[0].length;
        for(int i=0;i<ROWS;i++){
            for(int j=0;j<COLS;j++){
                if(grid[i][j] == '1'){
                    count+=1;
                    dfs(i,j,ROWS,COLS,grid);
                }
            }
        }

        return count;
    }

    private static void dfs(int r, int c, int ROWS, int COLS, char[][] grid){
        if((r>=0 && r<ROWS) && (c>=0 && c<COLS) && grid[r][c] == '1'){
            grid[r][c] = '0';
            dfs(r,c+1,ROWS,COLS,grid);
            dfs(r,c-1,ROWS,COLS,grid);
            dfs(r+1,c,ROWS,COLS,grid);
            dfs(r-1,c,ROWS,COLS,grid);
        }
    }
}